File: mobile/android/air/SubModules/WalletCore/src/main/java/app/twallet/air/walletcore/stores/ChainVisibilityStore.kt
package app.twallet.air.walletcore.stores

import app.twallet.air.walletcore.models.blockchain.MBlockchain
import app.twallet.air.walletcore.moshi.MApiTransaction
import org.json.JSONObject

/**
 * Tracks which networks are hidden in Settings → Networks.
 * Activity history (and related UI) should only cover enabled chains.
 */
object ChainVisibilityStore : IStore {
    @Volatile
    private var hiddenByNetwork: Map<String, Set<String>> = emptyMap()

    fun update(hiddenChainsByNetwork: Map<String, List<String>>) {
        hiddenByNetwork = hiddenChainsByNetwork.mapValues { (_, chains) -> chains.toSet() }
    }

    fun updateFromJson(json: JSONObject?) {
        if (json == null) {
            hiddenByNetwork = emptyMap()
            return
        }
        val parsed = mutableMapOf<String, Set<String>>()
        val keys = json.keys()
        while (keys.hasNext()) {
            val network = keys.next()
            val array = json.optJSONArray(network) ?: continue
            val chains = buildSet {
                for (i in 0 until array.length()) {
                    val chain = array.optString(i)
                    if (chain.isNotEmpty()) add(chain)
                }
            }
            parsed[network] = chains
        }
        hiddenByNetwork = parsed
    }

    fun isHidden(chain: String, network: String): Boolean {
        return hiddenByNetwork[network]?.contains(chain) == true
    }

    fun isActivityVisible(activity: MApiTransaction, network: String): Boolean {
        return activityChains(activity).all { !isHidden(it, network) }
    }

    private fun activityChains(activity: MApiTransaction): List<String> {
        return when (activity) {
            is MApiTransaction.Transaction -> listOfNotNull(chainForSlug(activity.slug))
            is MApiTransaction.Swap -> listOfNotNull(
                chainForSlug(activity.from),
                chainForSlug(activity.to),
            ).distinct()
        }
    }

    private fun chainForSlug(slug: String): String? {
        TokenStore.getToken(slug)?.chain?.let { return it }
        val parts = slug.split("-", limit = 2)
        if (parts.size == 2) {
            return MBlockchain.valueOfOrNull(parts[0])?.name
        }
        return MBlockchain.entries.firstOrNull { it.nativeSlug == slug }?.name
    }

    override fun wipeData() {
        hiddenByNetwork = emptyMap()
    }

    override fun clearCache() {
        // Visibility is persisted settings, not ephemeral cache
    }
}

File: mobile/ios/Air/SubModules/WalletCore/Stores/ChainVisibilityStore.swift
import Foundation
import WalletContext
import WalletCoreTypes

/// Tracks which networks are hidden in Settings → Networks.
/// Used to keep activity history (and other UI) scoped to enabled chains only.
public final class ChainVisibilityStore: @unchecked Sendable {
    public static let shared = ChainVisibilityStore()

    private let queue = DispatchQueue(label: "org.twallet.app.chain_visibility_store", attributes: .concurrent)
    private var hiddenByNetwork: [ApiNetwork: Set<ApiChain>] = [:]

    private init() {}

    public func update(hiddenChainsByNetwork: [String: [String]]) {
        var parsed: [ApiNetwork: Set<ApiChain>] = [:]
        for (networkRaw, chains) in hiddenChainsByNetwork {
            guard let network = ApiNetwork(rawValue: networkRaw) else { continue }
            parsed[network] = Set(chains.compactMap { ApiChain(rawValue: $0) })
        }
        queue.async(flags: .barrier) {
            self.hiddenByNetwork = parsed
        }
    }

    public func isHidden(_ chain: ApiChain, network: ApiNetwork) -> Bool {
        queue.sync {
            hiddenByNetwork[network]?.contains(chain) ?? false
        }
    }

    public func isActivityVisible(_ activity: ApiActivity, network: ApiNetwork) -> Bool {
        activityChains(activity).allSatisfy { !isHidden($0, network: network) }
    }

    private func activityChains(_ activity: ApiActivity) -> [ApiChain] {
        switch activity {
        case .transaction(let transaction):
            if let chain = getChainBySlug(transaction.slug) {
                return [chain]
            }
            return []
        case .swap(let swap):
            var chains: [ApiChain] = []
            if let from = getChainBySlug(swap.from) {
                chains.append(from)
            }
            if let to = getChainBySlug(swap.to), to != chains.first {
                chains.append(to)
            }
            return chains
        }
    }
}
